src/vmem/vmem.c

```diff
@@ -17,37 +17,87 @@ extern int __start_vmem, __stop_vmem;
     We therefore use __attribute__((weak)) so we can compile in the absence of these. */
 extern __attribute__((weak)) int __start_vmem, __stop_vmem;
 
+/**
+ * @brief VMEM Memory copy function - 32-bit version ONLY
+ * 
+ * This method is only capable of handling 32-bit source and destination
+ * addresses, which means that it CAN NOT handle the newest 64-bit addressing
+ * 
+ * @param to 32-bit destination address (virtual or physical)
+ * @param from 32-bit source address (virtual or physical)
+ * @param size Number fo bytes to copy
+ * @return void* 
+ */
 void * vmem_memcpy(void * to, const void * from, uint32_t size) {
 
+	return vmem_cpy((uint64_t)(uintptr_t)to, (uint64_t)(uintptr_t)from, (uint64_t)size);
+}
+
+/**
+ * @brief Write chunk of data to VMEM from physical memory to virtual memory
+ * 
+ * @param to Virtual address to write the data to
+ * @param from Physical address to read the data from
+ * @param size Number of bytes to transfer 
+ * @return void* 
+ */
+void * vmem_write(uint64_t to, const void * from, uint32_t size) {
+
+	for(vmem_t * vmem = (vmem_t *) &__start_vmem; vmem < (vmem_t *) &__stop_vmem; vmem++) {
+		/* Write to VMEM */
+		if ((to >= vmem->vaddr) && (to + (uint64_t)size <= vmem->vaddr + vmem->size)) {
+			printf("Write to vmem %s, to 0x%"PRIX64" from 0x%"PRIX32"\n", vmem->name, to, (uint32_t)from);
```

Comment:
We cannot have printfs for every read and write in production code :-)

Reply:
Nope, you are right - debug code - to be removed :-)